### ISSUE TITLE
When downloading YouTube live video, download from very beginning

### ISSUE THREAD

**Issue by fireattack** [ytdl-org/youtube-dl#21255]
## Checklist

- [x] I'm reporting a feature request
- [x] I've verified that I'm running youtube-dl version **2019.05.20**
- [x] I've searched the bugtracker for similar feature requests including closed ones

## Description

Currently, when downloading a (still) streaming live video on YouTube, youtube-dl will start from the current time.

I was wondering if we can add a way to download it from the very beginning instead. (I can confirm it is technical-wise possible at least, the video sources are available in m3u8 files).

This way, the user can get the full video once the stream is finished. Otherwise, at least the first few seconds will be missing.

Of course, you can start the download once the live is done typically to solve this problem. However, some uploaders delete their live stream immediately after premiere, makes it somewhat difficult. 

Thanks.

**Comment by guillaumedoriot2** [comment]
I don't have the same behavior: which option do you use? Aare you authenticated?

**Comment by makeworld-the-better-one** [comment]
I have downloaded an ongoing stream and it started from the beginning.

**Comment by fireattack** [comment]
I did a quick test myself again.

1. I use webcam to start a quick live stream (by using https://www.youtube.com/livestreaming/) at https://www.youtube.com/watch?v=12REdrcaP1w.
2. I waited for 1-2 min and then start to download it with `youtube-dl`.
3. I waited a few minutes, and then terminated my stream. 

The video file I downloaded is only 3min:

```
General
Complete name                            : D:\test-12REdrcaP1w.mp4
Format                                   : MPEG-4
Format profile                           : Base Media
Codec ID                                 : isom (isom/iso2/avc1/mp41)
File size                                : 13.2 MiB
Duration                                 : 3 min 0 s
Overall bit rate                         : 614 kb/s
Writing application                      : Lavf58.20.100
```

The stream, if recorded from very beginning, should be around 4:55 as you can see in above link.

**Comment by fireattack** [comment]
Also by simply looking at the network traffic, you can see youtube-dl will only fetch the _new_ video sections (instead of trying to catch all the available sections from the very beginning, which would have a much higher download speed).

**Comment by fireattack** [comment]
Another thing I want to mention is that if you stop the download manually by ctrl-C it will also generate a broken (merged) mp4 file. 

So you have to wait for the stream to end itself, which makes the testing hard.

**Comment by ealgase** [comment]
> Another thing I want to mention is that if you stop the download manually by ctrl-C it will generate a broken (merged) mp4 file. So you have to wait for the stream to end itself, which makes testing hard.
> 
> This doesn't seem to happen if you download a normal video from YouTube: you can always interrupt the download and the part file (after rename) can be played just fine.

Cancelling with CTRL+C works for me.

**Comment by fireattack** [comment]
> > Another thing I want to mention is that if you stop the download manually by ctrl-C it will generate a broken (merged) mp4 file. So you have to wait for the stream to end itself, which makes testing hard.
> > This doesn't seem to happen if you download a normal video from YouTube: you can always interrupt the download and the part file (after rename) can be played just fine.
> 
> Cancelling with CTRL+C works for me.

Did you try on a ongoing live stream? It works on normal video but not live stream here.

I just tried again with this (currently ongoing): https://www.youtube.com/watch?v=QWTNIDZw608

The result is the same. I'm on Windows 7 64 bit with newest FFMPEG.

**Comment by ealgase** [comment]
Try it on Linux?

**Comment by fireattack** [comment]
I don't use Linux, but I may give it a try later in VM.

**Comment by fireattack** [comment]
> Try it on Linux?

Hi, just tested in a Ubuntu 19 VM. Ctrl+C indeed doesn't generate broken file like on Windows (tested on Win 7 and Win 10).

But the main issue remains: it doesn't download from the very beginning on Linux either.

**Comment by ealgase** [comment]
Yeah, it doesn't download from the beginning, I was just troubleshooting the other issue.

**Comment by Bangaio65** [comment]
I don't think you can do it with just youtube-dl, but you can using ffmpeg.

Caveats: 
- as far as I know, youtube-dl can only grab the HLS (.m3u8) manifest. Some livestreams let you rewind up to 12 hours on the website (as far as I cant tell that is using the MPEG Dash (.mpd) manifest) but only 2 or 4 hours with the HLS stream. I don't think ffmpeg can correctly parse an .mpd manifest as of yet, at least last time I checked. And I don't know if youtube-dl can even get the DASH manifest.
- Some livestreams don't support rewinding

Instructions:
1. Grab the HLS manifest url using youtube-dl (add -f best if you want the best quality)
`youtube-dl --no-warnings -f "best" -g -- <youtube-url>`

2. Figure out how many seconds is each segment (it's usually 2 or 5) and convert seconds to segments.
If you're familiar with video editing, this is the time between keyframes. 
Say you want to go back 1 hour and the segment duration is 2 seconds. Then you do  (1 * 60 * 60) / 2 = 800. You need to "go back" 1800 segments on the stream.
If you just need to go back to the start of the stream, just input a large number on the next step.

3. Download using ffmpeg
add `-t <seconds or HH:MM:SS>` to specify a duration
it's recommended to save as `.ts` just in case the connection dies in the middle and you're left with an unusable `.mp4`
`ffmpeg -live_start_index -1800 -i <m3u8_url> -t 600 -c copy livestream01.ts`
or
`ffmpeg -live_start_index -1800 -i <m3u8_url> -t 600 -bsf:a aac_adtstoasc -c copy livestream01.mp4`

Hope this helps.

**Comment by nixklai** [comment]
I believe it is a ffmpeg-related issue, as I have marked the `-live_start_index` with an impossible number (-180000) and the output file still starts at the time of command execution.

For those whose debugging this issue: `--hls-use-mpegts --ouput "%(title)s-%(id)s.ts"` option helps turn the output into a `.ts` file for testing.

**Comment by nixklai** [comment]
The issue seems to be caused by the initial responses from YouTube are truncated to only 5 segments, like in this [test.m3u8 file](https://github.com/ytdl-org/youtube-dl/files/5830342/test.m3u8.txt) I download via cURL.


.

**Comment by johnhawkinson** [comment]
I had meant to comment last month and was distracted:

>  EDIT 2: It seems like an old problem that ffmpeg team refuses to fix

I don't think this is quite accurate, or at least, it is more nuanced than described. The bug report you cite, https://trac.ffmpeg.org/ticket/5811,  claims:

> ffmpeg  has an option to set the segment number where it must start reading.
...
>  But this option is totally broken, and ffmpeg always starts at the third last segment.

I concur with the analysis of 
>  * resolution:   => worksforme

That is, I use `-live_start_index N` with `ffmpeg` on a somewhat regular basis and it wors fine for me. I'm sure there's a reason @smalukav was having a problem, but it's not just broken all the time. I somehow doubt that's the failure with Youtube, but that's rank speculation on my part.

In any case, it doesn't seem like the `ffmpeg` developers are "refusing" — they report they cannot reproduce the problem and need a better reproduction case. That's a very different hing.
